pkg/kubelet/network/cni/cni.go

```diff
@@ -0,0 +1,232 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cni
+
+import (
+	"fmt"
+	"github.com/appc/cni/libcni"
+	cniTypes "github.com/appc/cni/pkg/types"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/kubelet/network"
+	kubeletTypes "k8s.io/kubernetes/pkg/kubelet/types"
+	"net"
+	"sort"
+)
+
+const (
+	CNIPluginName        = "cni"
+	DefaultPluginName    = "kubernetes-bridge"
+	DefaultNetDir        = "/etc/cni/net.d"
+	DefaultCNIDir        = "/opt/cni/bin"
+	VendorCNIDirTemplate = "/opt/%s/bin"
+)
+
+type cniNetworkPlugin struct {
+	cniNetworkMap map[string]*cniNetwork
+	host          network.Host
+}
+
+type cniNetwork struct {
+	name          string
+	NetworkConfig *libcni.NetworkConfig
+	CNIConfig     *libcni.CNIConfig
+}
+
+func ProbeNetworkPlugins() []network.NetworkPlugin {
+	configList := make([]network.NetworkPlugin, 0)
+	return append(configList, &cniNetworkPlugin{cniNetworkMap: getAllCNINetworks()})
+}
+
+func getAllCNINetworks() map[string]*cniNetwork {
+	networkMap := make(map[string]*cniNetwork)
+	files, err := libcni.ConfFiles(DefaultNetDir)
+	switch {
+	case err != nil:
+		return networkMap
+	case files == nil || len(files) == 0:
+		return networkMap
+	}
+
+	sort.Strings(files)
+	for _, confFile := range files {
+		conf, err := libcni.ConfFromFile(confFile)
+		if err != nil {
+			glog.Warningf("Error loading CNI config file %s: %v", confFile, err)
+			continue
+		}
+		// Search for vendor-specific plugins as well as default plugins in the CNI codebase.
```

Comment:
I think it is useful. The executable names may be clashing as @Symmetric pointed out, but also there could be a bunch of other files as dependencies. It will be wise to keep a directory per vendor as an option.